diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts b/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts
index bb10e6a480a2..539213387dfd 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts
@@ -60,6 +60,17 @@
 		enable-gpios = <&pio 5 6 GPIO_ACTIVE_HIGH>; /* LCD-BL-EN: PF6 */
 		power-supply = <&vdd_dsi>;
 	};
+
+	leds {
+		compatible = "gpio-leds";
+		
+		heartbeat_led {
+			label = "heartbeat_led";
+			gpios = <&pio 1 4 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+			linux,default-trigger = "heartbeat";
+		};
+	};
 };
 
 &ac_power_supply {
@@ -96,7 +107,7 @@
 		compatible = "bananapi,lhr050h41", "ilitek,ili9881c";
 		reg = <0>;
 		power-supply = <&vdd_dsi>;	/* VCC-LCD */
-		reset-gpios = <&pio 1 4 GPIO_ACTIVE_LOW>; /* LCD-RST: PD24 */
+		reset-gpios = <&pio 1 5 GPIO_ACTIVE_LOW>; /* LCD-RST: PD24 */
 		backlight = <&backlight>;
 	};
 };
