diff --git a/drivers/gpu/drm/panel/panel-ilitek-ili9881c.c b/drivers/gpu/drm/panel/panel-ilitek-ili9881c.c
index f7e528a38630..248551ddfafe 100644
--- a/drivers/gpu/drm/panel/panel-ilitek-ili9881c.c
+++ b/drivers/gpu/drm/panel/panel-ilitek-ili9881c.c
@@ -399,12 +399,12 @@ static int ili9881c_prepare(struct drm_panel *panel)
 		return ret;
 
 		
-	ili9881c_send_byte(0x2c);
+	ili9881c_send_byte(ctx, 0x2c);
 	for(int x=0;x<1280;x++)
 	{
-		for(y= 0;y<720;y++)
+		for(int y= 0;y<720;y++)
 		{
-			ili9881c_send_cmd_data(0x00, 0x00);
+			ili9881c_send_cmd_data(ctx, 0x00, 0x00);
 		}
 	}
 
